crates/biome_deserialize/src/impls.rs

```diff
@@ -0,0 +1,571 @@
+                diagnostics: &mut Vec<DeserializationDiagnostic>,
+            ) -> Option<Self::Output> {
+                Some(
+                    values
+                        .filter_map(|value| Deserializable::deserialize(value, diagnostics))
+                        .collect(),
+                )
+            }
+        }
+        value.deserialize(Visitor(PhantomData), diagnostics)
+    }
+}
+
+impl<T: Hash + Eq + Deserializable> Deserializable for IndexSet<T> {
+    fn deserialize(
+        value: impl DeserializableValue,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self> {
+        struct Visitor<T>(PhantomData<T>);
+        impl<T: Hash + Eq + Deserializable> DeserializationVisitor for Visitor<T> {
+            type Output = IndexSet<T>;
+            const EXPECTED_TYPE: ExpectedType = ExpectedType::ARRAY;
+            fn visit_array(
+                self,
+                values: impl Iterator<Item = impl DeserializableValue>,
+                _range: TextRange,
+                diagnostics: &mut Vec<DeserializationDiagnostic>,
+            ) -> Option<Self::Output> {
+                Some(
+                    values
+                        .filter_map(|value| Deserializable::deserialize(value, diagnostics))
+                        .collect(),
+                )
+            }
+        }
+        value.deserialize(Visitor(PhantomData), diagnostics)
+    }
+}
+
+impl<K: Hash + Eq + Deserializable, V: Deserializable, S: Default + BuildHasher> Deserializable
+    for HashMap<K, V, S>
+{
+    fn deserialize(
+        value: impl DeserializableValue,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self> {
+        struct Visitor<K, V, S>(PhantomData<(K, V, S)>);
+        impl<K: Hash + Eq + Deserializable, V: Deserializable, S: Default + BuildHasher>
+            DeserializationVisitor for Visitor<K, V, S>
+        {
+            type Output = HashMap<K, V, S>;
+            const EXPECTED_TYPE: ExpectedType = ExpectedType::MAP;
+            fn visit_map(
+                self,
+                members: impl Iterator<Item = (impl DeserializableValue, impl DeserializableValue)>,
+                _range: TextRange,
+                diagnostics: &mut Vec<DeserializationDiagnostic>,
+            ) -> Option<Self::Output> {
+                let mut result = Self::Output::default();
+                for (key, value) in members {
+                    let key = Deserializable::deserialize(key, diagnostics);
+                    let value = Deserializable::deserialize(value, diagnostics);
+                    if let (Some(key), Some(value)) = (key, value) {
+                        result.insert(key, value);
+                    }
+                }
+                Some(result)
+            }
+        }
+        value.deserialize(Visitor(PhantomData), diagnostics)
+    }
+}
+
+impl<K: Ord + Deserializable, V: Deserializable> Deserializable for BTreeMap<K, V> {
```

Comment:
We should not implement  `HashMap` and `BTreeMap`, both don't retain **insertion** order. We should instead implement `Deserializable` for `IndexMap`

Reply:
Bijective encoding is an important property to have, however it is more a concern for binary format. I don't see the value of guarantying this property for human-readable formats. We still can't guarantee bijective encoding for objects (and we won't, because a user shouldn't care about the order of object properties). Thus, I don't see the value of bothering for unordered maps and unordered array-like structure. Could you elaborate?

Reply:
Ordering **is important**! Two reasons:
- `biome.json#overrides` are applied in sequence from first to last in the order they are declared
- when we **serialize** `biome.json#overrides`, we want to keep the insertion order

Now, I understand why you want to apply this, and it's fine with me. However, it's also important to explain to our developers when using one type rather than another. For example, our `Configuration` uses only `StringSet` because it requires maintaining insertion order. So, as an action item, I believe we should explain to our contributors which type they should use to represents their data format (most of the time, a JSON).